src/System.Management.Automation/engine/NativeCommandProcessor.cs

```diff
@@ -528,95 +549,350 @@ internal override void Complete()
+                        else if (stream == MinishellStream.Debug ||
+                                 stream == MinishellStream.Verbose ||
+                                 stream == MinishellStream.Warning)
+                        {
+                            //Convert to string
+                            try
+                            {
+                                obj = LanguagePrimitives.ConvertTo(obj, typeof(string), CultureInfo.InvariantCulture);
+                            }
+                            catch (PSInvalidCastException)
                             {
-                                startPosition.Y = 0;
+                                continue;
                             }
+                        }
+                        result.Add(new ProcessOutputObject(obj, stream));
+                    }
+                }
+            }
+            catch (XmlException originalException)
+            {
+                string template = NativeCP.CliXmlError;
+                string message = string.Format(
+                    null,
+                    template,
+                    isOutput ? MinishellStream.Output : MinishellStream.Error,
+                    Path,
+                    originalException.Message);
+                XmlException newException = new XmlException(
+                    message,
+                    originalException);
+
+                ErrorRecord error = new ErrorRecord(
+                    newException,
+                    "ProcessStreamReader_CliXmlError",
+                    ErrorCategory.SyntaxError,
+                    Path);
+                result.Add(new ProcessOutputObject(error, MinishellStream.Error));
+            }
 
-                            Host.BufferCell[,] bufferContents = this.Command.Context.EngineHostInterface.UI.RawUI.GetBufferContents(
-                                new Host.Rectangle(startPosition, endPosition));
+            return result;
+        }
 
-                            StringBuilder lineContents = new StringBuilder();
-                            StringBuilder bufferText = new StringBuilder();
+        private void InitOutputQueue()
+        {
+            //if output is redirected, start reading output of process in queue.
+            if (_startInfo.RedirectStandardOutput || _startInfo.RedirectStandardError)
+            {
+                lock (_sync)
+                {
+                    if (!_stopped)
+                    {
+                        _nativeProcessOutputQueue = new ConcurrentQueue<ProcessOutputObject>();
 
-                            for (int row = 0; row < bufferContents.GetLength(0); row++)
+                        if (_startInfo.RedirectStandardOutput)
+                        {
+                            bool isFirstOutput = true;
+                            bool isXmlCliOutput = false;
+                            _nativeProcess.OutputDataReceived += new DataReceivedEventHandler((sender, e) =>
```

Comment:
What you are describing is #1908 and it's outside of the scope of this PR. I'd like to solve it with creating already piped processes and avoid expensive manual marshaling from stdout of one to stdin of another.